erpnext/buying/doctype/supplier/supplier.py

```diff
@@ -48,3 +48,74 @@ def on_trash(self):
 	def after_rename(self, olddn, newdn, merge=False):
 		if frappe.defaults.get_global_default('supp_master_name') == 'Supplier Name':
 			frappe.db.set(self, "supplier_name", newdn)
+
+def get_timeline_data(doctype, name):
+	'''returns timeline data based on linked records in dashboard'''
+	from six import iteritems
+	from frappe.utils import get_timestamp
+
+	out = {}
+
+	'''supplier quotation'''
+	items = dict(frappe.db.sql('''select transaction_date, count(*)
+		from `tabSupplier Quotation` where supplier=%s
+			and transaction_date > date_sub(curdate(), interval 1 year)
+			group by transaction_date''', name))
+
+	for date, count in items.iteritems():
+		timestamp = get_timestamp(date)
+		out.update({ timestamp: count })
+
+	'''purchase order'''
+	items = dict(frappe.db.sql('''select transaction_date, count(*)
+		from `tabPurchase Order` where supplier=%s
+			and transaction_date > date_sub(curdate(), interval 1 year)
+			group by transaction_date''', name))
+
+	for date, count in iteritems(items):
```

Comment:
Thanks for the contribution :) While the finer control and overriding in every Doctype is great, I'm slightly worried about code duplication; for example, this `for` functionality that takes in `items` and updates `out` could be commonified as a utility function `def update_timeline_data(items, out)` that is always called with the most recently updated `out`. There are 3-4 occurrence of these same 5 lines across 8 files, I'm sure they can be reduced :)

Reply:
Another immediate side effect is having to use a `for` to update `out`; there are 3-4 occurrence of these same 5 lines of the loop across 8 files.